src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -91,46 +98,53 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
             _packageInstaller = packageInstaller;
             _packageUninstaller = packageUninstaller;
             _packageSourceProvider = packageSourceProvider;
+            ResetPackageSources();
+
+            // Create a work queue to allow us to batch up flurries of notifications to be processed at some time in the
+            // future.  To prevent saturating the UI thread, we process no more than once every 250ms.
+            //
+            // Our work notifications are a simple pair that says if we should process the entire solution or, if not,
+            // which project we need to process.  Because it's just simple data, and the order doesn't matter, we can
+            // pass in an appropriate IEqualityComparer to dedupe notifications as they come in.
+            _workQueue = new AsyncBatchingWorkQueue<(bool changedSolution, ProjectId changedProject)>(
+                TimeSpan.FromMilliseconds(250),
+                ProcessBatchedChangesAsync,
+                EqualityComparer<(bool, ProjectId)>.Default,
+                asyncListener: null,
+                _cancellationTokenSource.Token);
         }
 
-        public ImmutableArray<PackageSource> GetPackageSources()
+        void IDisposable.Dispose()
+            => _cancellationTokenSource.Cancel();
+
+        private void ResetPackageSources()
+            => Interlocked.Exchange(ref _packageSources, new AsyncLazy<ImmutableArray<PackageSource>>(
+                c => ComputePackageSourcesAsync(c), cacheResult: true));
+
+        private async Task<ImmutableArray<PackageSource>> ComputePackageSourcesAsync(CancellationToken cancellationToken)
         {
-            // Only read from _packageSources once, since OnSourceProviderSourcesChanged could reset it to default at
-            // any time while this method is running.
-            var packageSources = _packageSources;
-            if (packageSources != null)
-            {
-                return packageSources;
-            }
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
```

Comment:
It's not clear to me if this is actually safe to use SwitchToMainThreadAsync() under an AsyncLazy since I don't imagine JTF can chase the dependency through the TaskCompletionSources that we make. This is very much a statement of "I know what I don't know" and the uncertainty there is worrisome to me.